### DISCUSSION TITLE
Prevent api function from sharing ressource with others

### DISCUSSION THREAD

**Discussion by BjoernRave** [vercel/next.js#13089] [Help]
I have one serverless function in the `/api` directory which uses puppeteer, making it barely usable on vercel. The problem is, that the serverless functions are sharing their ressources when deploying exceeding the 50mb limit for my other functions.

The puppeteer function is not sharing any functions or anything with the others.

Is there a way to isolate a serverless function?

This is the problematic function: 

```
import chrome from 'chrome-aws-lambda'
import { NextApiRequest, NextApiResponse } from 'next'
import puppeteer from 'puppeteer'
import { parse } from 'url'

export const generatePDF = async (url: string, token: string) => {
  const browser = await puppeteer.launch(
    process.env.NODE_ENV === 'production'
      ? {
          args: chrome.args,
          executablePath: await chrome.executablePath,
          headless: chrome.headless,
        }
      : {}
  )
  const page = await browser.newPage()
  await page.setViewport({
    width: 1500,
    height: 2121,
    deviceScaleFactor: 2,
  })

  await page.setCookie({
    name: 'authorization',
    value: token,
    url,
    expires: Date.now() / 1000 + 10,
  })
  await page.goto(url, { waitUntil: 'networkidle2' })
  const pdf = await page.pdf({
    printBackground: true,
    height: 1216,
    width: 860,
  })

  await browser.close()
  return pdf
}

export default async function (req: NextApiRequest, res: NextApiResponse) {
  const parsed = parse(req.url, true)

  if (!parsed.query.url || !req.headers.authorization) {
    return res.end('Please provide required infos')
  }

  try {
    const pdf = await generatePDF(
      decodeURIComponent(parsed.query.url as string),
      req.headers.authorization.replace('Bearer ', '')
    )
    res.end(pdf)
  } catch (error) {
    res.writeHead(500, { 'Content-Type': 'text/plain' })
    res.end('There was an error performing the action')
  }
}
```

**Comment by jensmeindertsma** [comment]
Hey @BjoernRave . If this is a big problem, perhaps you could try to create a new project on Vercel specifically for the API function. That way it doesn't need to share resources with the other code in your project. 50 MB sounds like a lot though, perhaps you should get in contact with [Vercel Support](https://vercel.com/support/request). I'm sure they can help you out with this.

**Reply by BjoernRave** [reply]
Hey @jensmeindertsma, thanks for the answer. Yes the size is due to puppeteer and is expected. Sure I could create a seperate project for it, but I would love to have it in the same to keep everything in the same place and to make development nicer

**Reply by jensmeindertsma** [reply]
Then I would suggest talking to the guys over at [Vercel](https://vercel.com/support/request), they know much more about their infrastructure than I do, and I'm sure they'll be able to help you with this.

**Comment by stnwk** [comment]
I found this in their documentation: https://vercel.com/docs/concepts/functions/introduction#bundling-serverless-functions

> In order to optimize resources, there is an internal process that bundles as many Serverless Functions as possible into a single Lambda. Such feature is currently only enabled for Next.js, but it will be enabled in other scenarios in the future.
> 
> This internal process shouldn't affect the developer in any case. **However, this behavior can be disabled in the project configuration via the functions property.**

But it doesn't say how exactly to achieve that... I've contacted vercel support about this already, let's see what they have to say :/
